src/mysql_protocol.erl

```diff
@@ -1220,6 +1214,96 @@ recv_packet(SockModule, Socket, Timeout, ExpectSeqNum, Acc) ->
 
+-spec send_file(module(), term(), Filename :: binary(), AllowedPaths :: [binary()],
+                SeqNum :: integer()) ->
+    {ok | {error, term()}, NextSeqNum :: integer()}.
+send_file(SockModule, Socket, Filename, AllowedPaths, SeqNum0) ->
+    {Result, SeqNum1} = case is_allowed_path(Filename, AllowedPaths) andalso
+                             file:open(Filename, [read, raw, binary]) of
+        false ->
+            {{error, not_allowed}, SeqNum0};
+        {ok, Handle} ->
+            {ok, SeqNum2} = send_file_chunk(SockModule, Socket, Handle, SeqNum0),
+            ok = file:close(Handle),
+            {ok, SeqNum2};
+        {error, _Reason} = E ->
+            {E, SeqNum0}
+    end,
+    {ok, SeqNum3} = send_packet(SockModule, Socket, <<>>, SeqNum1),
+    {Result, SeqNum3}.
+
+-spec is_allowed_path(file:name_all(), [file:name_all()]) -> boolean().
+is_allowed_path(Path, AllowedPaths) ->
+    Path1 = normalize_split_path(Path),
+    lists:any(
+        fun (AllowedPath) ->
+            AllowedPath1 = normalize_split_path(AllowedPath),
+            is_prefix_path(AllowedPath1, Path1)
+        end,
+        AllowedPaths
+    ).
+
+is_prefix_path([], _Path2) ->
+    true;
+is_prefix_path(_Path1, []) ->
+    false;
+is_prefix_path([Segment1|Path1], Path2=[Segment2|_]) when is_binary(Segment1), is_list(Segment2) ->
+    is_prefix_path([binary_to_list(Segment1)|Path1], Path2);
+is_prefix_path(Path1=[Segment1|_], [Segment2|Path2]) when is_list(Segment1), is_binary(Segment2) ->
+    is_prefix_path(Path1, [binary_to_list(Segment2)|Path2]);
+is_prefix_path([Segment1|Path1], [Segment2|Path2]) ->
+    case string:equal(Segment1, Segment2) of
+        true -> is_prefix_path(Path1, Path2);
+        false -> false
+    end.
+
+-spec normalize_split_path(file:name_all()) -> list().
+normalize_split_path(Path) ->
+    Path1 = filename:flatten(Path),
+    Path2 = case filename:pathtype(Path1) of
+        absolute ->
+            Path1;
+        relative ->
+            {ok, Cwd} = file:get_cwd(),
+            filename:absname_join(Cwd, Path1);
+        volumerelative ->
+            Volume = hd(filename:split(Path1)),
+            {ok, Cwd} = file:get_cwd(Volume),
+            filename:absname_join(Cwd, Path1)
+    end,
+    [Root|RelPath] = filename:split(Path2),
+    normalize_split_path(Root, RelPath, []).
+
+normalize_split_path(Root, [], Acc) ->
+    [Root|lists:reverse(Acc)];
+normalize_split_path(Root, ["."|More], Acc) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [<<".">>|More], Acc) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [".."|More], []) ->
+    normalize_split_path(Root, More, []);
+normalize_split_path(Root, [<<"..">>|More], []) ->
+    normalize_split_path(Root, More, []);
+normalize_split_path(Root, [".."|More], [_|Acc]) ->
+    normalize_split_path(Root, More, Acc);
+normalize_split_path(Root, [<<"..">>|More], [_|Acc]) ->
+    normalize_split_path(Root, More, Acc);
```

Comment:
Yes. As pointed out elsewhere, it may be possible to work on the real file system (like `filelib:safe_relative_path/2`, we need to get at the real file in the next step, anyway), but ignoring `..` is really just simpler, so we may end up with that, anyway.